llama.cpp

```diff
@@ -13909,7 +13909,7 @@ struct llama_context * llama_new_context_with_model(
     cparams.rope_freq_scale  = params.rope_freq_scale == 0.0f ? hparams.rope_freq_scale_train : params.rope_freq_scale;
 
     // this is necessary due to kv_self.n being padded later during inference
-    cparams.n_ctx = GGML_PAD(cparams.n_ctx, 32);
+    cparams.n_ctx = GGML_PAD(cparams.n_ctx, 256);
```

Comment:
I think we should be padding to the head size. For example, Phi-2 has HS=80 which is not a multiple of 256

Reply:
No, (for CUDA) it's better to pad to a power of 2 and the current CUDA implementation expects this padding to work correctly. The matrix shapes are VKQ = (D, BS) = V * KQ = (D, PAD) * (PAD, BS) = V * K * Q = (D, PAD) * (PAD, D) * (D, BS). Initially I did an implementation with PAD == D and warp counts scaling with D but that leads to issues with warp counts being e.g. 5 for head size 80. With the current implementation you get a better shape for KQ in terms of both the calculation and the softmax. For the calculation of VKQ awkward D values are treated via > 1 accumulators for VKQ parts. If the padding was to a multiple of D the kernel would need to deal with awkward shapes for all parts of the calculation instead of just a single part. There is also the consideration that exponential functions are very expensive and the smaller you choose D the more of them do you have to calculate.